### ISSUE TITLE
Text prompt

### ISSUE THREAD

**Issue by zzzyzh** [JayParanjape/biastuning#6]
I was under the impression that sam didn't provide a way to use text prompt. Does SAM release the training and testing code for text prompt?

Is your work based on SAM's own text prompt code?

Looking forward to your reply!

**Comment by JayParanjape** [comment]
Hi, 
You are right, SAM does not provide code for text prompt. But they mention in their paper that they use CLIP embeddings as the text prompt. Hence, in my code, I use pretrained CLIP to generate these embeddings and concatenate them to the sparse prompt embeddings of SAM. 
Hope that clears things

**Comment by zzzyzh** [comment]
Thank you for being so kind!

If I want to use biastuning on my dataset, do I need to generate organ text labels such as Liver and Kidney for each slice?

**Comment by JayParanjape** [comment]
During training, I am assuming you have annotated data per slice. So, you would need to generate a dataset class in Pytorch that creates training samples, with just the individual class pixels in the mask. For these, the text prompt would be the classname. So, in your example, each training point would be a slice, combined with the text prompt 'Kidney'/'Liver' and its corresponding ground truth. You can control how many images of each class you want to provide.

During testing, you just need the slice and the text prompt which you want for that slice

Hope that helps. Do let me know if you are still confused

**Comment by zzzyzh** [comment]
I get it!
Thank you very much for your help!

**Comment by zzzyzh** [comment]
I realised I missed a very crucial point, how is the data preprocessed? Let's say Kidney = 1 and Liver = 2. At this point a label contains both Kidney and Liver, do I need to reprocess this label as one label containing Kidney = 1 and one label containing Liver = 2?

**Comment by JayParanjape** [comment]
Correct, during training it expects the label to be split as you say. But it can be easily modified to use the label as is. For that, you will need to concatenate the text prompts in the correct way to sparse embeddings of SAM so that it views these as two different prompts for the same image. Then it will generate two masks for these two labels
